pinot-query-planner/src/test/java/org/apache/pinot/query/planner/logical/StagesTestBase.java

```diff
@@ -0,0 +1,232 @@
+import com.google.common.collect.ImmutableMap;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.query.planner.plannode.JoinNode;
+import org.apache.pinot.query.planner.plannode.MailboxReceiveNode;
+import org.apache.pinot.query.planner.plannode.MailboxSendNode;
+import org.apache.pinot.query.planner.plannode.PlanNode;
+import org.apache.pinot.query.planner.plannode.TableScanNode;
+import org.testng.annotations.AfterMethod;
+
+
+/**
+ * A base test class that can be used to write tests for stages using a fluent DSL.
+ *
+ * This class provides two features:
+ * <ul>
+ *   <li>Builders that can be used to create nodes in a fluent way.</li>
+ *   <li>Access to the stages that were created during the test with {@link #stage(int)}.</li>
+ * </ul>
+ *
+ * It is expected that each test method will call {@link #when(SimpleChildBuilder)} to create a new plan, which will
+ * populate the list of stages. After that, the test can look for the stages with the {@link #stage(int)} method and
+ * assert the expected behavior.
+ */
+public class StagesTestBase {
+  private final HashMap<Integer, MailboxSendNode> _stageRoots = new HashMap<>();
+
+  /**
+   * Clears the list of stages.
+   *
+   * This method is automatically called by the test framework, ensuring each test starts with a clean slate.
+   * This method can also be called in middle of the test, but that is not recommended given it usually means that the
+   * test is getting too complex and difficult to read and/or get insights from it in case of failure.
+   */
+  @AfterMethod
+  public void cleanup() {
+    _stageRoots.clear();
+  }
+
+  /**
+   * Creates a new plan that will have an initial stage.
+   *
+   * The stage will have a default {@link MailboxSendNode} whose stage will be 0 and its child the one created by the
+   * builder.
+   *
+   * Notice that this method does not offer any way to customize the initial send mailbox.
+   */
+  public MailboxSendNode when(SimpleChildBuilder<? extends PlanNode> builder) {
+    return sendMailbox(0, builder).build(0);
+  }
+
+  /**
+   * Returns a builder that can be used to create a new mailbox receive node.
+   *
+   * It is usually recommended to use {@link #exchange(int, SimpleChildBuilder)} instead of this method, given that
+   * {@code exchange} creates a pair of send and receive mailboxes and deals with the stageId management.
+   */
+  public SimpleChildBuilder<MailboxReceiveNode> receiveMailbox(SimpleChildBuilder<MailboxSendNode> childBuilder) {
+    return (stageId, mySchema, myHints) -> {
+      MailboxSendNode mailbox = childBuilder.build(stageId);
+      int nextStageId = mailbox.getStageId();
+      return new MailboxReceiveNode(stageId, mySchema, List.of(), nextStageId, null, null, null, null, false, false,
```

Comment:
```suggestion
      return new MailboxReceiveNode(stageId, mySchema, nextStageId, null, null, null, null, false, false,
```
This compilation error is causing the build to fail currently.